File: 01. DataStructures/04. Tree/03. AVL Tree/AVLTree.java
package treeDS;

import java.util.Scanner;

import treeDS.BST.Node;

public class AVLTree {

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		AVL_T at = new AVL_T();
		boolean flag = true;
		int val = 0;
		while (flag) {
			System.out.println();
			System.out.println("1 - Insert a Node in AVL Tree");
			System.out.println("2 - Print 2D");
			System.out.println("3 - isBalanced?");
			System.out.println("4 - Delete a Node in AVL Tree");
			System.out.println("0 - Exit");
			int ch = sc.nextInt();
			switch (ch) {
			case 1:
				System.out.print("Enter Value to Insert: ");
				val = sc.nextInt();
				at.root = at.insert(at.root, val);
				break;
			case 2:
				at.print2D(at.root, 5);
				break;
			case 3:
				at.height(at.root);
				System.out.println(at.isBal);
				break;
			case 4:
				System.out.print("Enter Value to Delete: ");
				val = sc.nextInt();
				at.root = at.delete(at.root, val);
				break;
			case 0:
				flag = false;
				break;
			default:
				System.out.println("Invalid choice");
			}
		}
	}
}

class AVL_T {
	class Node {
		int data;
		Node left;
		Node right;

		public Node(int val) {
			data = val;
			left = right = null;
		}
	}

	Node root;

	public AVL_T() {
		root = null;
	}

// ------ START Some Extra User Defined Functions ------

// START Get Balance factor of node N 
	public int getBalancedFac(Node r) {
		if (r == null)
			return -1;
		else
			return height(r.left) - height(r.right);
	}
// END Get Balance factor of node N 

// START Right Rotation
	public Node rightRotate(Node r) {
		Node nr = r.left; // new root
		Node T2 = nr.right;

		// perform right rotation
		nr.right = r;
		r.left = T2;

		return nr;
	}
// END Right Rotation

// START Left Rotation
	public Node leftRotate(Node r) {
		Node nr = r.right; // new root
		Node T2 = nr.left;

		// perform left rotation
		nr.left = r;
		r.right = T2;

		return nr;
	}
// END Left Rotation

// ------ END Some Extra User Defined Functions ------

// START 1 - Insert a Node in AVL Tree
	public Node insert(Node r, int key) {
		if (r == null) {
			return new Node(key);
		}

		// if the value is less than the root node then recur for the left subtree
		else if (key < r.data) {
			r.left = insert(r.left, key);
		}

		// else recur for the right subtree
		else if (key > r.data) {
			r.right = insert(r.right, key);
		}

		// key == root.data
		else {
			System.out.println("Value Already Exists!");
			return r;
		}

		int bf = getBalancedFac(r);
		// Left Case
		if (bf > 1 && key < r.left.data)
			return rightRotate(r);
		// Right Case
		else if (bf < -1 && key > r.right.data)
			return leftRotate(r);
		// Left-Right Case
		else if (bf > 1 && key > r.left.data) {
			r.left = leftRotate(r.left);
			return rightRotate(r);
		}
		// Right-Left Case
		else if (bf < -1 && key < r.right.data) {
			r.right = rightRotate(r.right);
			return leftRotate(r);
		}
		return r;
	}
// END 1 - Insert a Node in AVL Tree

// START 2 - Print 2D
	public void print2D(Node r, int space) {
		if (r == null) // base case
			return;

		print2D(r.right, space + 5);
		System.out.println();
		for (int i = 0; i < space; i++)
			System.out.print(" ");

		System.out.println(r.data);
		print2D(r.left, space + 5);
	}
// END 2 - Print 2D

// START Height of Tree
// START 3 - isBalanced?
	static boolean isBal = true; // only to check isBalanced?

	public static int height(Node r) {
		if (r == null)
			return -1;

		else {
			int lheight = height(r.left);
			int rheight = height(r.right);

			// only to check isBalanced?, ignore this condition for finding the height
			if (Math.abs(lheight - rheight) > 1) {
				isBal = false;
			}
			//
			return Math.max(lheight, rheight) + 1;
		}
	}
// END 3 - isBalanced?
// END Height of Tree

// START 4 Delete a Node in AVL Tree
	// function to find the maximum value from the left-subtree
	public static int maxLeft(Node node) {
		if (node.right != null)
			return maxLeft(node.right);
		else
			return node.data;
	}
	
	public Node delete(Node r, int val) {
		if (r == null)
			return null;
		if (val < r.data) {
			r.left = delete(r.left, val);
		} else if (val > r.data) {
			r.right = delete(r.right, val);
		} 
		// when the node to be deleted is found
		else {
			if (r.left != null && r.right != null) {
				int lmax = maxLeft(r.left);
				r.data = lmax;
				r.left = delete(r.left, lmax);
			} 
			// if the node has only a left child
			else if (r.left != null) {
				return r.left;
			} 
			// if the node has only a right child
			else if (r.right != null) {
				return r.right;
			} 
			// if the node has no child i.e., it is a leaf node
			else {
				return null;
			}
		}

		int bf = getBalancedFac(r);
		
		// if bf is 2, it means that the deletion will happen in the right subtree
		// this will result in L-L-Imbalance or L-R-Imbalance
		/*
		  => bf(r.left)=1 means remaining structure in this is:-
		 		30 bf=2
		 	   /
		 	  20   bf=1
		 	 /
		 	10     bf=0
		
		 => bf(r.left)=0 means remaining structure in this is:-
		 	   30 bf=2
		 	  /
		 	 20   bf=0
		 	/  \
		   10  25   bf=0 for both
	  */
		if (bf == 2 && getBalancedFac(r.left) >= 0)
			return rightRotate(r);
		else if (bf == 2 && getBalancedFac(r.left) == -1) {
			r.left = leftRotate(r.left);
			return rightRotate(r);
		}
		
		// if bf is -2, it means that the deletion will happen in the left subtree
		// this will result in R-R-Imbalance or R-L-Imbalance
		/*
		   => bf(r.right)=-1 means remaining structure in this is:-
		 		10 bf=-2
		 		  \
		 		  20 bf=-1
		 		    \
		 		    30 bf=0
		 
		
		
		  => bf(r.right)=0 means remaining structure in this is:-
		 	   10  bf=-2
		 		 \
		 		 20 bf=0
		 		/  \
		 	  15    30 bf=0 for both
	  */
		else if (bf == -2 && getBalancedFac(r.right) <= 0)
			return leftRotate(r);
		else if (bf == -2 && getBalancedFac(r.right) == 1) {
			r.right = rightRotate(r.right);
			return leftRotate(r);
		}
		return r;
	}
// END 4 Delete a Node in AVL Tree
}

File: 01. DataStructures/04. Tree/03. AVL Tree/AVLTree.js

class Node {
    constructor(val) {
        this.val = val;
        this.left = null;
        this.right = null;
    }
}

class avlTree {

   constructor(data) {
       this.root = new Node(data);
       this.globleSpaceForPrinting = 5;
   } 

   height(root) {

    let maxHeight = 0;
    return goRecursive(root, 0);

    function goRecursive(root, curruntHeight) {
        if(!root) return 0;

        curruntHeight += 1;
        maxHeight = Math.max(maxHeight, curruntHeight);
        goRecursive(root.left, curruntHeight);
        goRecursive(root.right, curruntHeight);
        return maxHeight;
    }
   }

   rightRotate(root) {
    // console.log(root);
    let temp = root.left;
    let temp1 = temp.right;

    temp.right = root;
    root.left = temp1;
    return temp;
   }

   leftRotate(root) {
    let temp = root.right;
    let temp1 = temp.left;

    temp.left = root;
    root.right = temp1;

    return temp;
   }

   getBalanceFactor(root) {
    if(!root) {
        return -1;
    } else {
        return this.height(root.left) - this.height(root.right);
    }
    }

    insertNode(root, data) {

        const that = this;        
        this.root = goRecursive(root, data);
        function goRecursive(root, data) {
            if(!root) {
                return new Node(data);
            } else if (data < root.val) {
                root.left = goRecursive(root.left, data);
            } else if (data > root.val) {
                root.right = goRecursive(root.right, data);
            } else {
                return 'can\'t insert duplicates';
            }
    
            const bf = that.getBalanceFactor(root);
            if(bf > 1 && data < root.left.val) {
                // dod right balance;
               return that.rightRotate(root);
            } else if (bf < -1 && data > root.right.val) {
               return  that.leftRotate(root);
                // do left balance;
            } else if (bf > 1 && data > root.left.val) {
                root.left = that.leftRotate(root.left);
               return that.rightRotate(root); 
                // do left rotate 
                // do  right rotate 
            } else if(bf < -1 && data < root.right.val) {
                root.right = that.rightRotate(root.right);
               return  that.leftRotate(root);
                // do right rotate
                // do left rotate;
            }
            return root;    
        }

        return this.root;
    }
    leftM(root) {
        if(root.right) {
            return this.leftM(root.right);
        } else {
            return root.val;
        }
    }
    deleteNode(root, val) {
        const that = this;
        this.root = goRecursive(root, val);

        function goRecursive(root, val) {

            if(root && val > root.val) {
                // go right
                root.right = goRecursive(root.right, val);
            } else if(root && val < root.val) {
                root.left = goRecursive(root.left, val);
                // go left
            } else if(root && val === root.val) {

               if(root.left && root.right) {
                   // node has left and right child
                   const leftMax = that.leftM(root.left);
                   root.val = leftMax;
                   root.left = that.deleteNode(root.left, leftMax);
               } else if(root.left !== null) {
                   // node has only left child
                   return root.left;
               } else if(root.right !== null) {
                   // node has only right child
                   return root.right;
               } else {
                //    it's a leaf node delete it
                root = null;
               }

            } else {
                return false;
            }

            const bf = that.getBalanceFactor(root);
           
            if(bf == 2 && that.getBalanceFactor(root.left) > 0) {
                return that.rightRotate(root);
            } else if(bf == -2 && that.getBalanceFactor(root.right) < 0) {
                return that.leftRotate(root);
            } else if(bf == 2 && that.getBalanceFactor(root.left) < 0) {
                root.left = that.leftRotate(root.left);
                return that.rightRotate(root);
            } else if(bf == -2 && that.getBalanceFactor(root.right) > 0) {
                root.right = that.rightRotate(root.right);
                return that.leftRotate(root);
            }

            return root;
        }

        return this.root;
    }
    print2D(root, space) {
        if(!root) {
            return;
        }
        
        space = this.globleSpaceForPrinting + space;
        this.print2D(root.right, space);

        let curruntSpace = '';
        for(let i = this.globleSpaceForPrinting; i < space; i++) {
            curruntSpace += " ";
        }
        console.log(curruntSpace + root.val + '\n');
        this.print2D(root.left, space);
    }
}

const myAvl = new avlTree(60);
// console.log(myAvl.root);

myAvl.insertNode(myAvl.root,50);
myAvl.insertNode(myAvl.root,40);
myAvl.insertNode(myAvl.root,65);
myAvl.insertNode(myAvl.root,82);
myAvl.insertNode(myAvl.root,12);
myAvl.insertNode(myAvl.root,42);
myAvl.insertNode(myAvl.root,30);
// myAvl.deleteNode(myAvl.root, 42);
myAvl.deleteNode(myAvl.root, 50);
// myAvl.deleteNode(myAvl.root, 42);
// myAvl.deleteNode(myAvl.root, 60);
// myAvl.deleteNode(myAvl.root, 82);
// myAvl.deleteNode(myAvl.root, 65);
// myAvl.deleteNode(myAvl.root, 42);
// myAvl.deleteNode(myAvl.root, 42);
// myAvl.deleteNode(myAvl.root, 42);
// myAvl.deleteNode(myAvl.root, 42);
// myAvl.deleteNode(myAvl.root, 40);
myAvl.print2D(myAvl.root, 5);
